mat/dense.go

```diff
@@ -67,11 +67,33 @@ func NewDense(r, c int, data []float64) *Dense {
 	}
 }
 
-// reuseAs resizes an empty matrix to a r×c matrix,
-// or checks that a non-empty matrix is r×c.
+// ReuseAs changes the receiver if it IsZero() to be of size r×c.
+//
+// ReuseAs re-uses the backing data slice if it has sufficient capacity,
+// otherwise a new slice is allocated. The data is then zeroed.
+//
+// ReuseAs panics if the receiver is not zero-sized, and panics if
+// the input sizes are less than one. To zero a matrix for re-use,
+// Reset should be used.
```

Comment:
```
//
// ReuseAs is not safe to use if the receiver shares backing data with
// another Matrix.
```

Reply:
Isn't this obvious from the behavior? The same is true with `Reset`, `Augment`, `Copy`, `CloneFrom`, etc. Any operation that can change the backing slice is not safe to use when sharing backing data.

Reply:
It is also true for those. I think that given that people mess up `append` and this is more complex, it's worth noting. Probably in the package doc as well that these are possible. The issue is not about shared data though, it's about potentially clearing data that's not obviously shared.

Reply:
I mean, it is about shared data, in the sense that if you don't share data there are no problems. There also aren't any problems if you aren't tricky about it (splitting rows and columns of matrices, which is actually reasonably hard to do with the package).

I don't think this function itself presents problems. The dangerous one is `Reset`, which enables `ReuseAs` and 85% of the potential for misuse of shared data (i.e. reshaping receivers and destination matrices).

Reply:
The second paragraph is the issue, yes. Perhaps the warning should be on reset. There the warning would be that by resetting matrix guards against leaky sharing are lost.